urld558a3d988_gao715108023_resourceMonitor_tgz-pJ8-com_sinopec_agent_AgentJ8/src/com/sinopec/agent/rpc/DynamicInfoClient.java

```diff
@@ -0,0 +1,138 @@
+package com.sinopec.agent.rpc;
+
+import com.sinopec.agent.share.SharedArea;
+import com.sinopec.dao.DynamicInfoInterface;
+import com.sinopec.test.PrintTestInfo;
+import com.sinopec.utils.LogUtils;
+import com.sinopec.utils.TimeUtil;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.net.MalformedURLException;
+import java.rmi.Naming;
+import java.rmi.NotBoundException;
+import java.rmi.Remote;
+import java.rmi.RemoteException;
+
+/**
+ * Created with IntelliJ IDEA. User: Administrator Date: 13-10-23 Time: 下午2:58
+ * To change this template use File | Settings | File Templates.
+ */
+public class DynamicInfoClient implements Runnable {
+
+	private static final Log LOG = LogFactory.getLog(DynamicInfoClient.class);
+
+	private static volatile boolean running = true;
+
+	private static volatile boolean service = true;
+
+	private int samplingInterval;
+
+	private String curGroupIP;
+
+	private int port;
+
+	private Remote remote;
+
+	private int totalRetryCount;
+
+	private int retryConnTimeout = 3000;
+
+	private String nextGroupIP;
+
+	public DynamicInfoClient(int samplingInterval, String curGroupIP, int port, int timeoutConnToNextGroup) {
+		this.samplingInterval = samplingInterval;
+		this.curGroupIP = curGroupIP;
+		this.port = port;
+		this.totalRetryCount = (timeoutConnToNextGroup * 60 * 1000) / retryConnTimeout;
+	}
+
+	public boolean isServerExists(String bindName) {
+		boolean result = false;
+		try {
+			remote = Naming.lookup(bindName);
+			if (remote != null)
+				result = true;
+		} catch (NotBoundException e) {
+			// e.printStackTrace(); //To change body of catch statement use File
+			// | Settings | File Templates.
+		} catch (MalformedURLException e) {
+			// e.printStackTrace(); //To change body of catch statement use File
+			// | Settings | File Templates.
+		} catch (RemoteException e) {
+			// e.printStackTrace(); //To change body of catch statement use File
+			// | Settings | File Templates.
+		}
+		return result;
+	}
+
+	public void connectToGroup() {
+		int retryCount = 1;
+		boolean b = true;
+		while (running) {
+			if (isServerExists("rmi://" + curGroupIP + ":" + port + "/heartbeat")) {
+				running = false;
+				service = true;
+			} else {
+				LOG.info("Retrying connect to Group: " + curGroupIP + ".Already tries " + retryCount + " time(s).");
+				if (retryCount >= totalRetryCount && b) {
+					curGroupIP = getNextGroupIP();
+					retryCount = 1;
+					b = false;
+				}
+				retryCount++;
+				try {
+					Thread.sleep(retryConnTimeout);
+				} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)